crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+        &mut self,
+        new_best_hash: Block::Hash,
+    ) -> Result<(), sp_blockchain::Error> {
+        let current_best_hash = match self.best_hash() {
+            Some(h) => h,
+            // Collect bundle from the last K blocks if it have not process any block
+            None => {
+                self.collect_bundle_from_last_k(new_best_hash)?;
+                return Ok(());
+            }
+        };
+
+        let route = sp_blockchain::tree_route(&*self.client, current_best_hash, new_best_hash)?;
+        let (retracted, enacted) = (route.retracted(), route.enacted());
+
+        // Remove bundles from the stale fork
+        for retracted_block in retracted {
+            match self.bundle_stored_in_last_k.front() {
+                Some((block_hash, _)) if *block_hash == retracted_block.hash => {
+                    self.bundle_stored_in_last_k.pop_front();
+                }
+                bb => {
+                    return Err(sp_blockchain::Error::Application(Box::from(
+                        format!(
+                            "Got wrong block from the bundle-collector, expect {:?}, got {:?}, this should not happen",
+                            retracted_block,
+                            bb.map(|(block_hash, _)| block_hash),
+                        ),
+                    )));
+                }
+            }
+        }
+
+        // Add bundles from the new block of the best fork
+        for enacted_block in enacted {
+            let block_bundles = self.collect_bundle_from_block(enacted_block.hash)?;
+            self.bundle_stored_in_last_k.push_front(block_bundles);
+        }
+
+        // Remove blocks from the back end to keep at most the bundle of the last K blocks
+        self.bundle_stored_in_last_k.truncate(self.confirm_depth_k);
+
+        Ok(())
+    }
+
+    pub async fn update_recent_stored_bundles(&mut self, new_best_hash: Block::Hash) {
+        let pre_best_hash = self.best_hash();
+        if let Err(err) = self.on_import_new_best_block(new_best_hash) {
+            tracing::error!(
+                %err,
+                "Failed to handle new best block for bundle-collector"
+            );
+        }
+        if pre_best_hash != self.best_hash() {
+            self.bundle_syncer
+                .update(self.bundle_stored_in_last_k.clone());
+        }
+    }
+}
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
+    #[allow(clippy::type_complexity)]
+    bundle_stored_in_last_k: Arc<Mutex<VecDeque<(Block::Hash, HashSet<Hash>)>>>,
+}
+
+impl<Block: BlockT> BundleStoredInLastK<Block> {
+    fn new() -> Self {
+        BundleStoredInLastK {
+            bundle_stored_in_last_k: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
+        *self.bundle_stored_in_last_k.lock() = incoming;
```

Comment:
Pls consider doing `a std::mem::swap()` with an empty `VecDequeue<>` under lock, so that the actual free/heap operations happen outside the lock

Reply:
1) It is not even a micro, it is a nano-optimization
2) Are you sure that'll be the case? I didn't verify this particular case, but in most other cases Rust actually drops things in reverse order, so it might end up being the same, but will hurt readability quite a bit.

Reply:
yeah it only helps if the the VecDequeue being freed is big enough. With k = 100, and lot of transactions per block, this would be helpful.

And I'm sure. Because the assignment would trigger the dtor of the previous VecDequeue, while lock is  still held